src/lambdas/createUser/createUser.js

```diff
@@ -0,0 +1,68 @@
+const DynamoDB = require('aws-sdk/clients/dynamodb');
+const db = new DynamoDB.DocumentClient();
+const Log = require('@dazn/lambda-powertools-logger');
+const cleanDeep = require('clean-deep');
+const TABLE_NAME = process.env.TABLE_NAME || '';
+
+module.exports.handler = async (event, context) => {
+    try {
+        // get data from event.body
+        if (!event.body) throw 'no body received';
+        const item = typeof event.body === 'object' ? event.body : JSON.parse(event.body);
+
+        // Get UUID for user id from the lambda request id
+        const userId = context.awsRequestId;
+
+        // current time
+        const timeNow = new Date().toISOString();
+
+        // extract for composite state#city#zip
+        const { city, state, zip } = item.address;
+        const scz = `${state}#${city}#${zip}`.toLowerCase().replace(/\s/g, '');
+
+        // system computed values
+        const computedValues = {
+            pk: userId,
+            sk: `USER`,
+            createdAt: timeNow,
+            updatedAt: timeNow,
+            scz,
+        };
+
+        // spread computed values over passed in values
+        const combinedItems = { ...item, ...computedValues };
+
+        // store the new user item
+        const params = {
+            TableName: TABLE_NAME,
+            Item: cleanDeep(combinedItems),
+            ConditionExpression: 'attribute_not_exists(sk)', // disallow overwrite
+        };
+        await db.put(params).promise();
+
+        // cloudwatch log
+        Log.info('User Created', { userId });
+
+        // respond success
+        return httpResponse(201, { id: userId });
+    } catch (err) {
+        // dynamo did not store because the record already exists
+        if (err.code === 'ConditionalCheckFailedException') {
+            return httpResponse(409, 'a user with this ID already exists');
```

Comment:
Can you please share your thoughts on the pros/cons of the approach you've taken here for checking if the user already exists.

Especially when weighted against doing an initial check against the table.

Reply:
In DynamoDB, when doing a put to an existing primary key, it will overwrite an existing item, which may not be the desired effect.  By adding the conditional check, verifying that a particular attribute does not exist, it blocks the overwrite.   My approach to using the conditional write was to avoid having to first do a 'get' to read any potential existing item to see if it exists, and then having to follow it up with a 'put'.  

Thank you for calling attention to this section as it made me realize my check was against the sort key 'sk' and should have been against the partition key 'pk'.  Earlier in the design I was storing the UUID as the sort key, but then changed to store the UUID as the partition key instead and forgot to update the condition expression.  This code should be updated so that the conditional check is against the partition key: 'attribute_not_exists(pk)'.